### PR TITLE
[bf16] Add dtype-bf16 supporting with naive_amp

### PR DESCRIPTION
**Pull request by ht-zhou**
Add dtype-bf16 supporting with naive_amp.

### CODE REVIEW

File: tests/test_amp/test_naive_bf16.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+def run_dist(rank, world_size, port):
+    colossalai.launch(config=dict(), rank=rank, world_size=world_size, port=port, host='localhost')
+    run_naive_bf16_amp()
+
+
+@pytest.mark.dist
+@rerun_if_address_is_in_use()
+def test_naive_bf16_amp():
+    world_size = 1
```
Comment (feifeibear) [inline tests/test_amp/test_naive_bf16.py:87]: also test world_size=2?

File: tests/test_amp/test_naive_bf16.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    # create layer
+    test_models = ['repeated_computed_layers', 'nested_model', 'resnet18']
+    for test_name in test_models:
+        get_component_func = non_distributed_component_funcs.get_callable(test_name)
+        model_builder, train_dataloader, _, optim_class, _ = get_component_func()
+
+        # create model
+        naive_fp16_amp_model = model_builder(checkpoint=True).cuda()
+        naive_bf16_amp_model = copy.deepcopy(naive_fp16_amp_model).cuda().bfloat16()
```
Comment (feifeibear) [inline tests/test_amp/test_naive_bf16.py:39]: naive_fp16_amp_model.clone()

File: tests/test_amp/test_naive_bf16.py (hunk 1 above)
Comment (ht-zhou) [reply tests/test_amp/test_naive_bf16.py:87]: yes, and it also works

File: tests/test_amp/test_naive_bf16.py (hunk 2 above)
Comment (ht-zhou) [reply tests/test_amp/test_naive_bf16.py:39]: it refers to test_naive_fp16.py
and 'NetWithRepeatedlyComputedLayers' object has no attribute 'clone'

**Review by feifeibear — APPROVED** [review]

**Comment by FrankLeeeee** [comment]
Hi, @ht-zhou , the CI suggest that some tests failed due to your code change. Can you fix them?

File: colossalai/amp/__init__.py [hunk 3]
```diff
@@ -13,14 +13,15 @@
 __all__ = ['convert_to_amp', 'convert_to_naive_amp', 'convert_to_apex_amp', 'convert_to_torch_amp', 'AMP_TYPE']
 
 
-def convert_to_amp(model: nn.Module, optimizer: Optimizer, criterion: _Loss, mode: AMP_TYPE, amp_config: Config = None):
+def convert_to_amp(model: nn.Module, optimizer: Optimizer, criterion: _Loss, mode: AMP_TYPE, use_bf16: bool = False, amp_config: Config = None):
```
Comment (1SAA) [inline colossalai/amp/__init__.py:16]: passing a bool paramter use_bf16 is not a good way. could you passing a precision type? you can use .to(torch.bfloat16) or .to(torch.float16) to control your target precision type.

File: colossalai/testing/comparison.py [hunk 4]
```diff
@@ -13,7 +13,7 @@ def assert_not_equal(a: Tensor, b: Tensor):
 
 
 def assert_close(a: Tensor, b: Tensor, rtol: float = 1e-5, atol: float = 1e-8):
-    assert torch.allclose(a, b, rtol=rtol, atol=atol), f'expected a and b to be close but they are not, {a} vs {b}'
+    assert torch.allclose(a.cuda().float(), b.cuda().float(), rtol=rtol, atol=atol), f'expected a and b to be close but they are not, {a} vs {b}'
```
Comment (1SAA) [inline colossalai/testing/comparison.py:16]: why you move all tensors to cuda? can you add some parameters in this function to control this?

File: colossalai/amp/__init__.py [hunk 5]
```diff
@@ -13,14 +14,15 @@
 __all__ = ['convert_to_amp', 'convert_to_naive_amp', 'convert_to_apex_amp', 'convert_to_torch_amp', 'AMP_TYPE']
 
 
-def convert_to_amp(model: nn.Module, optimizer: Optimizer, criterion: _Loss, mode: AMP_TYPE, amp_config: Config = None):
+def convert_to_amp(model: nn.Module, optimizer: Optimizer, criterion: _Loss, mode: AMP_TYPE, precision_type: dtype, amp_config: Config = None):
```
Comment (1SAA) [inline colossalai/amp/__init__.py:17]: can you move precision_type into amp_config?

File: colossalai/amp/__init__.py [hunk 6]
```diff
@@ -47,6 +49,6 @@ def convert_to_amp(model: nn.Module, optimizer: Optimizer, criterion: _Loss, mod
     elif mode == AMP_TYPE.APEX:
```
Comment (1SAA) [inline colossalai/amp/__init__.py:49]: you need to check whether precision is bf16 here.

File: colossalai/amp/naive_amp/__init__.py [hunk 7]
```diff
@@ -1,20 +1,24 @@
...
+from torch import dtype
+import torch
 from colossalai.utils import is_no_pp_or_last_stage
 from .naive_amp import NaiveAMPOptimizer, NaiveAMPModel
 from .grad_scaler import DynamicGradScaler, ConstantGradScaler
 from ._fp16_optimizer import FP16Optimizer
 
 
-def convert_to_naive_amp(model: nn.Module, optimizer: Optimizer, amp_config):
+def convert_to_naive_amp(model: nn.Module, optimizer: Optimizer, amp_config, precision_type: dtype = torch.float16):
```
Comment (1SAA) [inline colossalai/amp/naive_amp/__init__.py:13]: move precision_type to amp_config.

File: colossalai/amp/naive_amp/__init__.py [hunk 8]
```diff
@@ -51,6 +55,8 @@ def convert_to_naive_amp(model: nn.Module, optimizer: Optimizer, amp_config):
             kwargs[param.name] = amp_config.pop(param.name)
     grad_scaler = scaler_class(**kwargs)
     optimizer = NaiveAMPOptimizer(optimizer, grad_scaler, **amp_config)
+    if precision_type==torch.bfloat16:
```
Comment (1SAA) [inline colossalai/amp/naive_amp/__init__.py:58]: why we need to change it again?

File: colossalai/testing/comparison.py [hunk 9]
```diff
@@ -13,7 +13,7 @@ def assert_not_equal(a: Tensor, b: Tensor):
 
 
 def assert_close(a: Tensor, b: Tensor, rtol: float = 1e-5, atol: float = 1e-8):
-    assert torch.allclose(a, b, rtol=rtol, atol=atol), f'expected a and b to be close but they are not, {a} vs {b}'
+    assert torch.allclose(a.to(torch.float), b.to(torch.float), rtol=rtol, atol=atol), f'expected a and b to be close but they are not, {a} vs {b}'
```
Comment (1SAA) [inline colossalai/testing/comparison.py:16]: you need to add two flags for this function. one is `same_type`, if it is true, you turn them into the same type. other is `same_device`, if it is true, you turn them into the same device.

**Comment by binmakeswell** [comment]
Hi @ht-zhou Is this PR still active?

**Pull request closed without merging**